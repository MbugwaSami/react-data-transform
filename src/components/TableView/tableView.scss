.rdt-table-view {
    border: 1px solid #dee2e6;
    border-radius: 10px;
    height: 100%;
    table {
        font-family: arial, sans-serif;
        border-collapse: collapse;
        width: 100%;

        tbody {
            tr:nth-child(odd) {
                background-color: #0000000d;
            }
        }

        td,
        th {
            border-bottom: 1px solid #dee2e6;
            text-align: left;
            padding: 8px;
        }
    }
}