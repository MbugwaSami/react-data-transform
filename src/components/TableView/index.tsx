import { Fragment } from "react";

import "./tableView.scss";

const TableView = <T extends { [key: string]: any }>(
  props: TableViewProps<T>
) => {
  const { colums, tableData, tableName } = props;

  const RenderColums = () => {
    return (
      <tr className="rdt-table-head">
        {colums?.map((column) => {
          return <th key={column.dataKey}>{column.title}</th>;
        })}
      </tr>
    );
  };

  const RenderRows = () => {
    return (
      <Fragment>
        {tableData?.map((row) => {
          return (
            <tr key={row?.id} className="rdt-table-row">
              {colums?.map((column) => {
                return (
                  <td key={column.dataKey + row?.id}>{row[column.dataKey]}</td>
                );
              })}
            </tr>
          );
        })}
      </Fragment>
    );
  };

  return (
    <div className="rdt-table-view">
      {!tableData?.length ? (
        <div className="rdt-empty-table">{tableName} has no data </div>
      ) : (
        <table className="rdt-table">
          <thead>
            <RenderColums />
          </thead>
          <tbody>
            <RenderRows />
          </tbody>
        </table>
      )}
    </div>
  );
};

export default TableView;
