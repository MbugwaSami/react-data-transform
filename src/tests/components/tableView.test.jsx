import { mount } from "enzyme";
import TableView from "../../components/TableView";
import { tableDataMock, tableColumnsMock } from "../mockData";

describe("Table View ", () => {
  let wrapper;
  const tableProps = {
    tableData: tableDataMock,
    colums: tableColumnsMock,
  };
  beforeEach(() => {
    wrapper = mount(<TableView {...tableProps} />);
  });
  it("should render <TabView />", () => {
    expect(wrapper).toMatchSnapshot();
  });
  it("should render a table with all rows", () => {
    expect(wrapper.find(".rdt-table").length).toBe(1);
    expect(wrapper.find(".rdt-table-head").length).toBe(1);
    expect(wrapper.find(".rdt-table-row").length).toBe(tableDataMock.length);
  });

  it("should render a table with all columns", () => {
    expect(wrapper.find(".rdt-table").length).toBe(1);
    expect(wrapper.find(".rdt-table-head").children().length).toBe(
      tableColumnsMock.length
    );
    expect(wrapper.find(".rdt-table-row").at(1).children().length).toBe(
      tableColumnsMock.length
    );
  });

  it("should render a div with a message when there is no data ", () => {
    wrapper.setProps({ tableData: [] });
    expect(wrapper.find(".rdt-table").length).toBe(0);
    expect(wrapper.find("div.rdt-empty-table").length).toBe(1);
  });
});
