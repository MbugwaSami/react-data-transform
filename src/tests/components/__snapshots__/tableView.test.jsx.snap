// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table View  should render <TabView /> 1`] = `
<TableView
  colums={
    Array [
      Object {
        "dataKey": "c1",
        "title": "C1",
      },
      Object {
        "dataKey": "c2",
        "title": "C2",
      },
      Object {
        "dataKey": "c3",
        "title": "C3",
      },
      Object {
        "dataKey": "c4",
        "title": "C4",
      },
      Object {
        "dataKey": "c5",
        "title": "C5",
      },
    ]
  }
  tableData={
    Array [
      Object {
        "c1": "1",
        "c2": "1",
        "c3": "1",
        "c4": "1",
        "c5": "1",
        "id": 1,
      },
      Object {
        "c1": "1",
        "c2": "1",
        "c3": "1",
        "c4": "1",
        "c5": "1",
        "id": 2,
      },
      Object {
        "c1": "1",
        "c2": "1",
        "c3": "1",
        "c4": "1",
        "c5": "1",
        "id": 3,
      },
      Object {
        "c1": "1",
        "c2": "1",
        "c3": "1",
        "c4": "1",
        "c5": "1",
        "id": 4,
      },
    ]
  }
>
  <div
    className="rdt-table-view"
  >
    <table
      className="rdt-table"
    >
      <thead>
        <RenderColums>
          <tr
            className="rdt-table-head"
          >
            <th
              key="c1"
            >
              C1
            </th>
            <th
              key="c2"
            >
              C2
            </th>
            <th
              key="c3"
            >
              C3
            </th>
            <th
              key="c4"
            >
              C4
            </th>
            <th
              key="c5"
            >
              C5
            </th>
          </tr>
        </RenderColums>
      </thead>
      <tbody>
        <RenderRows>
          <tr
            className="rdt-table-row"
            key="1"
          >
            <td
              key="c11"
            >
              1
            </td>
            <td
              key="c21"
            >
              1
            </td>
            <td
              key="c31"
            >
              1
            </td>
            <td
              key="c41"
            >
              1
            </td>
            <td
              key="c51"
            >
              1
            </td>
          </tr>
          <tr
            className="rdt-table-row"
            key="2"
          >
            <td
              key="c12"
            >
              1
            </td>
            <td
              key="c22"
            >
              1
            </td>
            <td
              key="c32"
            >
              1
            </td>
            <td
              key="c42"
            >
              1
            </td>
            <td
              key="c52"
            >
              1
            </td>
          </tr>
          <tr
            className="rdt-table-row"
            key="3"
          >
            <td
              key="c13"
            >
              1
            </td>
            <td
              key="c23"
            >
              1
            </td>
            <td
              key="c33"
            >
              1
            </td>
            <td
              key="c43"
            >
              1
            </td>
            <td
              key="c53"
            >
              1
            </td>
          </tr>
          <tr
            className="rdt-table-row"
            key="4"
          >
            <td
              key="c14"
            >
              1
            </td>
            <td
              key="c24"
            >
              1
            </td>
            <td
              key="c34"
            >
              1
            </td>
            <td
              key="c44"
            >
              1
            </td>
            <td
              key="c54"
            >
              1
            </td>
          </tr>
        </RenderRows>
      </tbody>
    </table>
  </div>
</TableView>
`;
